### PR TITLE
CBOR: Basic Support for Tags

### PR DESCRIPTION
**Pull request by JesusMcCloud**
Introduces *basic* support for CBOR tags.

Basic means that no semantic checks are performed yet, so you freely use reserved tags (e.g. serialize a cornucopia of fruit and other produce using tags reserved for dates, URIs, MIME messages, negative bignums and so forth) and no error will be thrown.

* New annotations `ValueTags` and `KeyTags` to specify tags for keys and values, respectively
  * Tags can be nested freely as specified in the annotations
  * By default, tags will be validated upon deserialization. If no annotations are present on a property, validation is omitted. Behaviour can be customized (see below).
  * By default, tags will be written as annotated.  Behaviour can be customized (see below).
* Introduces the following configuration parameters  to the `Cbor` class (names are self-explanatory):
    * `writeKeyTags`
    * `writeValueTags`
    * `verifyKeyTags`
    * `verifyValueTags`

This is the first PR in a series of planned enhancements, but we still consider it valuable on its own (as it also affects #1560).

In the long run, we plan to support semantically correct serialization for all reserved CBOR tags. This includes separate contributions to koltinx.datetime and the native support for self-describing CBOR. The latter of which is planned to enable the deserialization of CBOR structures, while also enabling access to the original underlying raw byte-string. This is a crucial features for conveniently working with (i.e. verifying) signed data. (Which, in turn, is a requirement for a 100% pure Kotlin multiplatform COSE implementation.)<br> Hence, we want to use this opportunity to also collect early feedback from those who are more familiar with kotlinx.serialization than we are.

### CODE REVIEW

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *     val b: ByteArray  // CBOR major type 6 1234567(major type 4: an array of data items).
+ * )
+ * ```
+ *
+ * See [RFC 7049 2.4. Optional Tagging of Items](https://datatracker.ietf.org/doc/html/rfc7049#section-2.4).
+ */
+@SerialInfo
+@Target(AnnotationTarget.PROPERTY)
+@ExperimentalSerializationApi
+public annotation class ValueTags(@OptIn(ExperimentalUnsignedTypes::class) vararg val tags: ULong) {
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: Here, IDEA says `The opt-in annotation is redundant: no matching experimental API is used ` and also when using the annotation in a client project, IDEA suggests to add `@OptIn(ExperimentalUnsignedTypes::class)` when using `@ValueTags`. Is there something off?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * Specifies that a property shall be tagged and serialized as CBOR major type 6: optional semantic tagging
+ * of other major types.
+ * For types other than [ByteArray], [ByteString] will have no effect.
+ *
+ * Example usage:
+ *
+ * ```
+ * @Serializable
+ * data class Data(
+ *     @Tagged(1337uL)
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: Description (`@Tagged`) does not match actual name `@ValueTags`

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.*
+
+/**
+ * Specifies that a property shall be tagged and serialized as CBOR major type 6: optional semantic tagging
+ * of other major types.
+ * For types other than [ByteArray], [ByteString] will have no effect.
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: That's not really correct for this annotation, is it?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ *     @Tagged(1337uL)
+ *     @ByteString
+ *     val a: ByteArray, // CBOR major type 6 1337(major type 2: a byte string).
+ *
+ *     @Tagged(1234567uL)
+ *     val b: ByteArray  // CBOR major type 6 1234567(major type 4: an array of data items).
+ * )
+ * ```
+ *
+ * See [RFC 7049 2.4. Optional Tagging of Items](https://datatracker.ietf.org/doc/html/rfc7049#section-2.4).
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: Better link to the newer RFC at https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        public const val REGEX: ULong = 35u;
+        public const val MIME_MESSAGE: ULong = 36u;
+        public const val CBOR_SELF_DESCRIBE: ULong = 55799u;
+    }
+}
+
+@SerialInfo
+@Target(AnnotationTarget.PROPERTY)
+@ExperimentalSerializationApi
+public annotation class KeyTags(@OptIn(ExperimentalUnsignedTypes::class) vararg val tags: ULong)
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: Missing kdoc

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt [hunk 6]
```diff
@@ -21,32 +21,32 @@ class CborReaderTest {
     @Test
     fun testDecodeIntegers() {
         withDecoder("0C1903E8") {
-            assertEquals(12L, nextNumber())
-            assertEquals(1000L, nextNumber())
+            assertEquals(12L, nextNumber(null))
```
Comment (nodh) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt]: Why not leave the optional parameter value out?

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt [hunk 7]
```diff
@@ -659,6 +705,32 @@ class CborReaderTest {
         }
     }
 
+    @Test
+    fun testVerifyTagsOnStrings() {
+        /*
+         * 84                                # array(4)
```
Comment (nodh) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt]: Formatting of the comment `# array(4)` is slightly off

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt [hunk 8]
```diff
@@ -14,15 +14,15 @@ data class Simple(val a: String)
...
-        val str: String,
-        val i: Int,
-        val nullable: Double?,
-        val list: List<String>,
-        val map: Map<Int, Boolean>,
-        val inner: Simple,
-        val innersList: List<Simple>,
-        @ByteString val byteString: ByteArray,
-        val byteArray: ByteArray
+    val str: String,
```
Comment (nodh) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt]: Changing the whitespace of existing code should be left out imho

**Review by nodh — CHANGES_REQUESTED** [review]
Works great, but small changes requested

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt [hunk 9]
```diff
@@ -364,23 +411,30 @@ internal class CborDecoder(private val input: ByteArrayInput) {
...
         readByte()
         return arr
     }
 
-    fun nextString(): String {
-        skipOverTags()
+    fun nextString(tag: ULong) = nextString(ulongArrayOf(tag))
+    fun nextString(tags: ULongArray? = null) = nextTaggedString(tags).first
+
+    //used to r
```
Comment (nodh) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt]: used to r?

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt (hunk 1 above)
Comment (JesusMcCloud) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: This seems to be an IDEA hiccup. without the annotation, it does not compile.

In addition: I'm curious about comments from upstream maintainers and I suspect the type could change to something other than `ULong`

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt (hunk 2 above)
Comment (JesusMcCloud) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: true, fixed!

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt (hunk 6 above)
Comment (JesusMcCloud) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt]: thanks! those are indeed remnants of an early stage of development

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt (hunk 5 above)
Comment (JesusMcCloud) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: added

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt (hunk 8 above)
Comment (JesusMcCloud) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt]: reverted

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt (hunk 9 above)
Comment (JesusMcCloud) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoding.kt]: fixed!

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt (hunk 7 above)
Comment (JesusMcCloud) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborReaderTest.kt]: fixed

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt (hunk 3 above)
Comment (JesusMcCloud) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/ValueTags.kt]: right, thanks! I also renamed the File itself to Tags.kt

**Comment by JesusMcCloud** [comment]
submitting to upstream

**Pull request closed without merging**